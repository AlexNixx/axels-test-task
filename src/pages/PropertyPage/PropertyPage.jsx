import { useEffect } from 'react';
import { Link, useParams } from 'react-router-dom';
import { Carousel, Stack } from 'react-bootstrap';

import {
    StyledButton,
    StyledContainer,
    StyledImage
} from './PropertyPageStyles';

import { convertToUSAPrice } from '../../utils/price/convertToUSAPrice';
import { mockData } from '../../mockData/mockData';

export const PropertyPage = () => {
    const { id } = useParams();

    const mockItem = mockData.filter((item) => item.id === Number(id))[0];

    useEffect(() => {
        window.scrollTo({ top: 0, left: 0, behavior: 'smooth' });
    }, []);

    return (
        <StyledContainer>
            <Stack gap={2}>
                <Link to={'/'}>
                    <StyledButton variant="dark">Back</StyledButton>
                </Link>

                <Carousel>
                    {mockItem.image.map((src) => (
                        <Carousel.Item key={src}>
                            <StyledImage
                                src={src}
                                alt={`${mockItem.title}-photo`}
                            />
                        </Carousel.Item>
                    ))}
                </Carousel>

                <h2>{mockItem.title}</h2>
                <h5>Seller: {mockItem.seller}</h5>
                <h6>Price: {convertToUSAPrice(mockItem.price)}</h6>
                <p>{mockItem.description}</p>
            </Stack>
        </StyledContainer>
    );
};
