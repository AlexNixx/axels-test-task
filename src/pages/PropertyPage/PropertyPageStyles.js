import styled from "styled-components";
import {Container, Image, Button} from "react-bootstrap";

export const StyledContainer = styled(Container)`
  padding: 2rem;
`

export  const StyledImage = styled(Image)`
  width: 100%;
  height: 40rem;
  overflow: hidden;
`

export const StyledButton = styled(Button)`
  width: 10rem;
`