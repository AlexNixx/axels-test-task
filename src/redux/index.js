import { applyMiddleware, combineReducers, createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';

import propertiesReducer from './ducks/properties';
import propertyReducer from './ducks/property';

import { rootWatcher } from './rootWatcher';

const sagaMiddleware = createSagaMiddleware();

const rootReducer = combineReducers({
    propertiesReducer,
    propertyReducer
});

export const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));

sagaMiddleware.run(rootWatcher);
