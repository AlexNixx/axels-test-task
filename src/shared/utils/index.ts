export * from './test/renderWithProviders';
export * from './test/showLocation';
export * from './convertPrice/convertPrice';
