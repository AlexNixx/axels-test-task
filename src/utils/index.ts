export * from './test/renderWithProviders';
export * from './test/showLocation';
export * from './test/mockProperties';
export * from './convertPrice/convertPrice';
