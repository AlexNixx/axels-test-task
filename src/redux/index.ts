import { applyMiddleware, combineReducers, createStore } from 'redux';
import { TypedUseSelectorHook, useSelector, useDispatch } from 'react-redux';
import createSagaMiddleware from 'redux-saga';

import propertiesReducer from './ducks/properties';
import propertyReducer from './ducks/property';
import { rootWatcher } from './rootWatcher';

const sagaMiddleware = createSagaMiddleware();

const rootReducer = combineReducers({
    propertiesReducer,
    propertyReducer
});

export const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));

sagaMiddleware.run(rootWatcher);

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
