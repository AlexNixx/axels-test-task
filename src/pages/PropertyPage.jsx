import { useEffect } from 'react';
import { Carousel, Col, Row } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useParams } from 'react-router-dom';

import { fetchProperty } from '../redux/ducks/property';
import { convertPrice } from '../shared/utils';
import {
    StyledButton,
    StyledContainer,
    StyledImage
} from '../styles/pages/PropertyPageStyles';

export const PropertyPage = () => {
    const { id } = useParams();

    const dispatch = useDispatch();
    const property = useSelector(state => state.propertyReducer.property);
    const error = useSelector(state => state.propertyReducer.error);
    const loading = useSelector(state => state.propertyReducer.loading);

    useEffect(() => {
        dispatch(fetchProperty(id));
    }, [dispatch, id]);

    if (loading) return <p>Loading...</p>;

    if (error) return <p>{error}</p>;

    return (
        <StyledContainer>
            <Link to='/'>
                <StyledButton variant='dark'>Back</StyledButton>
            </Link>
            <Row>
                <Col xs={12} lg={7}>
                    <Carousel>
                        {property?.images.map(src => (
                            <Carousel.Item key={src}>
                                <StyledImage
                                    src={src}
                                    alt={`${property?.title}-photo`}
                                />
                            </Carousel.Item>
                        ))}
                    </Carousel>
                </Col>
                <Col>
                    <h2>{property?.title}</h2>
                    <h5>Seller: {property?.seller}</h5>
                    <h6>Price: {convertPrice(property?.price)}</h6>
                    <p>{property?.description}</p>
                </Col>
            </Row>
        </StyledContainer>
    );
};

PropertyPage.propTypes = {};
