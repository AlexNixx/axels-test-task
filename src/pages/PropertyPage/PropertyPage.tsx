import { useEffect } from 'react';
import { Carousel, Col, Row } from 'react-bootstrap';
import { Link, useParams } from 'react-router-dom';

import { fetchProperty } from '../../redux/ducks/property';
import { useAppDispatch, useAppSelector } from '../../redux';
import { convertPrice } from '../../shared/utils';
import {
    StyledButton,
    StyledContainer,
    StyledImage
} from '../../styles/pages/PropertyPageStyles';

export const PropertyPage = () => {
    const { id } = useParams();

    const dispatch = useAppDispatch();
    const property = useAppSelector(state => state.propertyReducer.property);
    const error = useAppSelector(state => state.propertyReducer.error);
    const loading = useAppSelector(state => state.propertyReducer.loading);

    useEffect(() => {
        dispatch(fetchProperty(Number(id)));
    }, [dispatch, id]);

    if (loading) return <p>Loading...</p>;

    if (error) return <p>{error}</p>;

    return (
        <StyledContainer data-testid='propertyPage'>
            <Link to='/'>
                <StyledButton variant='dark'>Back</StyledButton>
            </Link>
            <Row>
                <Col xs={12} lg={7}>
                    <Carousel>
                        {property?.images.map(src => (
                            <Carousel.Item key={src}>
                                <StyledImage
                                    src={src}
                                    alt={`${property?.title}-photo`}
                                />
                            </Carousel.Item>
                        ))}
                    </Carousel>
                </Col>
                <Col>
                    <h2>{property?.title}</h2>
                    <h5>Seller: {property?.seller}</h5>
                    <h6>Price: {convertPrice(property?.price)}</h6>
                    <p>{property?.description}</p>
                </Col>
            </Row>
        </StyledContainer>
    );
};
