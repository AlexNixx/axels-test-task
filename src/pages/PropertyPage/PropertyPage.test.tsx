import { screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { PropertyPage } from './PropertyPage';

import { renderWithProviders } from '../../shared/utils';
import {
    fetchProperty,
    setError,
    setProperty
} from '../../redux/ducks/property';
import { store } from '../../redux';

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useDispatch: () => mockDispatch
}));

describe('PropertyPage', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    test('displays loading state', () => {
        store.dispatch(fetchProperty(1));

        renderWithProviders(<PropertyPage />, {});

        expect(mockDispatch).toHaveBeenCalledTimes(1);
        expect(screen.getByTestId('loadingText')).toBeInTheDocument();
    });

    test('displays error state', async () => {
        const errorMessage = 'An error occurred';

        store.dispatch(setError(errorMessage));
        renderWithProviders(<PropertyPage />, {});

        expect(mockDispatch).toHaveBeenCalledTimes(1);
        expect(screen.getByTestId('errorText')).toBeInTheDocument();
        expect(screen.getByTestId('errorText')).toContainHTML(errorMessage);
    });

    test('renders property', async () => {
        store.dispatch(
            setProperty({
                id: 1,
                title: 'Test Property',
                price: 100,
                address: '123 Test St',
                seller: 'Test Seller',
                description: 'Test Description',
                images: ['test.jpg']
            })
        );
        const { container } = renderWithProviders(<PropertyPage />, {});

        expect(screen.getByText(/Test Property/)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
