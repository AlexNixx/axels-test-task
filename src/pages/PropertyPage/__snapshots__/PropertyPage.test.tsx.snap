// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Property Details Page should be display properties details 1`] = `
<div>
  <div
    class="sc-bcPKhP ftJLmf container"
    data-testid="propertyPage"
  >
    <a
      href="/"
    >
      <button
        class="sc-dkjKgF lYdyv btn btn-dark"
        type="button"
      >
        Back
      </button>
    </a>
    <div
      class="row"
    >
      <div
        class="col-lg-7 col-12"
      >
        <div
          class="carousel slide"
        >
          <div
            class="carousel-indicators"
          >
            <button
              aria-current="true"
              aria-label="Slide 1"
              class="active"
              data-bs-target=""
              type="button"
            />
          </div>
          <div
            class="carousel-inner"
          >
            <div
              class="active carousel-item"
            >
              <img
                alt="Test Property-photo"
                class="sc-grXZZQ eZoUFJ"
                src="test.jpg"
              />
            </div>
          </div>
          <a
            class="carousel-control-prev"
            href="#"
            role="button"
            tabindex="0"
          >
            <span
              aria-hidden="true"
              class="carousel-control-prev-icon"
            />
            <span
              class="visually-hidden"
            >
              Previous
            </span>
          </a>
          <a
            class="carousel-control-next"
            href="#"
            role="button"
            tabindex="0"
          >
            <span
              aria-hidden="true"
              class="carousel-control-next-icon"
            />
            <span
              class="visually-hidden"
            >
              Next
            </span>
          </a>
        </div>
      </div>
      <div
        class="col"
      >
        <h2
          data-testid="propertyTitle"
        >
          Test Property
        </h2>
        <h5
          data-testid="propertySeller"
        >
          Seller: 
          Test Seller
        </h5>
        <h6
          data-testid="propertyPrice"
        >
          Price: 
          $100.00
        </h6>
        <p
          data-testid="propertyDescription"
        >
          Test Description
        </p>
      </div>
    </div>
  </div>
</div>
`;
